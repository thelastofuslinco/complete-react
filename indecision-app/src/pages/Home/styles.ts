import styled from 'styled-components'

export const HomeContainer = styled.div`
  display: flex;
  flex-direction: column;
  height: 100vh;
  gap: 1rem;
`
