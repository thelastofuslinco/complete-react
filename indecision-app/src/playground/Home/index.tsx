import React from 'react'
import Header from '../Header'
import List from '../List'
import Form from '../Form'

interface Props {}

interface State {
  options: Array<string>
  selectedOption: number
}

class Home extends React.Component<Props, State> {
  constructor(props) {
    super(props)
    this.handleMakeDecision = this.handleMakeDecision.bind(this)

    this.state = {
      options: [],
      selectedOption: null
    }
  }

  handleMakeDecision() {
    const randomNum = Math.random() * this.state.options.length

    this.setState((prevValue) => ({
      ...prevValue,
      selectedOption: Math.floor(randomNum)
    }))
  }

  render(): React.ReactNode {
    return (
      <div>
        <Header />
        <button
          disabled={!this.state.options.length}
          onClick={this.handleMakeDecision}
        >
          What should i do?
        </button>

        <Form
          onSubmit={async (value) => {
            const response = await new Promise<{
              username: string
            }>((resolve) =>
              setTimeout(() => {
                resolve(value)
              })
            )

            this.setState((prevValue) => ({
              ...prevValue,
              options: [...prevValue.options, response.username]
            }))
          }}
        />
        <List value={this.state.selectedOption} options={this.state.options} />
      </div>
    )
  }
}

export default Home
