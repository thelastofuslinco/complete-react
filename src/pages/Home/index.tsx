import { Component } from 'react'
import Header from '../../components/Header'
import Options from './components/Options'
import SendMessage from './components/SendMessage'
import OptionModal from './components/OptionModal'

interface Props {}

interface State {
  options: Array<{ id: string; value: string; createdAt: string }>
  selectedOption: string
  showModal: boolean
}

class Home extends Component<Props, State> {
  state = {
    options: [],
    selectedOption: null,
    showModal: false
  }

  handleMakeDecision = () => {
    const randomNum = Math.floor(Math.random() * this.state.options.length)

    this.setState((prevState) => ({
      selectedOption: prevState.options[randomNum].value,
      showModal: true
    }))
  }

  handleDeleteOption = async (id: string) => {
    const response_id = await new Promise<string>((resolve) =>
      setTimeout(() => {
        resolve(id)
      }, 1500)
    )

    this.setState((prevValue) => ({
      options: prevValue.options.filter((option) => option.id !== response_id)
    }))
  }

  handleAddOption = async (value: string) => {
    const response = await new Promise<string>((resolve, reject) =>
      setTimeout(() => {
        if (value === '') {
          reject(new Error('Write some message'))
        }
        resolve(value)
      })
    )

    this.setState((prevValue) => ({
      options: [
        ...prevValue.options,
        {
          id: crypto.randomUUID(),
          value: response,
          createdAt: new Date().toISOString()
        }
      ]
    }))
  }

  handleDeleteALLOption = () => {
    this.setState({
      options: []
    })
  }

  render() {
    return (
      <main className="pageContainer">
        <Header
          title="Indecision app"
          subtitle="Put your life in the hands of a computer"
        />
        <div className="widget">
          <button
            className="big_button"
            disabled={!this.state.options.length}
            onClick={this.handleMakeDecision}
          >
            What should i do?
          </button>
          <div className="widget_header">
            <span>Your options</span>
            <button
              className="link_button"
              onClick={this.handleDeleteALLOption}
            >
              Remove all
            </button>
          </div>
          <Options
            onDelete={this.handleDeleteOption}
            options={this.state.options}
          />
          <SendMessage onSubmit={this.handleAddOption} />
          {this.state.showModal && (
            <OptionModal
              option={this.state.selectedOption}
              onClose={() => this.setState({ showModal: false })}
            />
          )}
        </div>
      </main>
    )
  }
}

export default Home
