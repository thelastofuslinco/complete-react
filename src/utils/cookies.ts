export function setCookie(name, value, days) {
  const d = new Date()
  d.setTime(d.getTime() + days * 24 * 60 * 60 * 1000)
  const expires = 'expires=' + d.toUTCString()
  document.cookie = `${name}=${value};${expires};path=/;Secure;HttpOnly;SameSite=Strict`
}

export function getCookie(cookieName) {
  const name = cookieName + '='
  const ca = document.cookie.split(';')

  for (let i = 0; i < ca.length; i++) {
    let c = ca[i]
    while (c.charAt(0) == ' ') {
      c = c.substring(1)
    }
    if (c.indexOf(name) == 0) {
      return c.substring(name.length, c.length)
    }
  }
  return ''
}
