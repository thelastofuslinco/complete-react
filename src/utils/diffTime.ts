export const diffTime = (time) => {
  const date1 = new Date(time).getTime()
  const date2 = new Date().getTime()
  const diff = Math.abs(date2 - date1)
  const diffTime = {
    s: Math.floor(diff / 1000),
    m: Math.floor(diff / (1000 * 60)),
    h: Math.floor(diff / (1000 * 60 * 60))
  }

  if (diffTime.h > 1) {
    return `${diffTime.h} hour ago`
  } else if (diffTime.m > 1) {
    return `${diffTime.m} minutes ago`
  } else {
    return `${diffTime.s} seconds ago`
  }
}
