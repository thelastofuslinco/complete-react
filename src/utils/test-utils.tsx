import { render } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import expensesSlice from '../store/slices/expensesSlice'

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { expenses: expensesSlice },
      preloadedState
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions })
  }
}
