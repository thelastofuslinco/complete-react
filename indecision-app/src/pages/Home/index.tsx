import React from 'react'
import Header from '../../components/Header'
import Options from '../../components/Options'
import Form from '../../components/Form'
import { HomeContainer } from './styles'
import { v4 } from 'uuid'

interface Props {}

interface State {
  options: Array<{ id: string; value: string }>
  selectedOption: string
}

class Home extends React.Component<Props, State> {
  constructor(props) {
    super(props)
    this.handleMakeDecision = this.handleMakeDecision.bind(this)
    this.handleAddOption = this.handleAddOption.bind(this)
    this.handleDeleteOption = this.handleDeleteOption.bind(this)
    this.state = {
      options: [],
      selectedOption: null
    }
  }

  handleMakeDecision() {
    const randomNum = Math.floor(
      Math.random() * this.state.options.length
    ).toString()

    this.setState((prevState) => ({
      selectedOption: prevState.options[randomNum].id
    }))
  }

  async handleDeleteOption(id: string) {
    const response_id = await new Promise<string>((resolve) =>
      setTimeout(() => {
        resolve(id)
      }, 1500)
    )
    console.log(response_id)

    this.setState((prevValue) => ({
      options: prevValue.options.filter((option) => option.id !== response_id)
    }))
  }

  async handleAddOption(value: string) {
    const response = await new Promise<string>((resolve, reject) =>
      setTimeout(() => {
        if (value === '') {
          reject(new Error('Write some message'))
        }
        resolve(value)
      })
    )

    this.setState((prevValue) => ({
      options: [
        ...prevValue.options,
        {
          id: v4(),
          value: response
        }
      ]
    }))
  }

  componentDidUpdate(): void {
    console.log(this.state)
  }

  render() {
    return (
      <HomeContainer>
        <Header
          title="Indecision app"
          subtitle="Put your life in the hands of a computer"
        />
        <button
          disabled={!this.state.options.length}
          onClick={this.handleMakeDecision}
        >
          What should i do?
        </button>
        <Options
          onDelete={this.handleDeleteOption}
          value={this.state.selectedOption}
          options={this.state.options}
        />
        <Form onSubmit={this.handleAddOption} />
      </HomeContainer>
    )
  }
}

export default Home
