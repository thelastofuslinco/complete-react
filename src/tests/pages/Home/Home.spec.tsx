import { screen } from '@testing-library/react'
import { renderWithProviders } from '../../../utils/test-utils'
import Home from '../../../pages/Home'

describe('Home page', () => {
  test('should render page', () => {
    renderWithProviders(<Home />)
    const element = screen.getByRole('main')
    expect(element).toMatchSnapshot()
  })
})
