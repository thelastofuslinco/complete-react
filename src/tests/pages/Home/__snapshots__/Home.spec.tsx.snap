// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page should render page 1`] = `
<main
  class="pageContainer"
>
  <header
    class="header_container"
  >
    <h1>
      Indecision app
    </h1>
    <span>
      Put your life in the hands of a computer
    </span>
  </header>
  <div
    class="widget"
  >
    <button
      class="big_button"
      disabled=""
    >
      What should i do?
    </button>
    <div
      class="widget_header"
    >
      <span>
        Your options
      </span>
      <button
        class="link_button"
      >
        Remove all
      </button>
    </div>
    <div
      class="optionsContainer"
    />
    <form
      class="form_container"
    >
      <input
        name="text"
      />
      <button
        type="submit"
      >
        send
      </button>
    </form>
  </div>
</main>
`;
